tests/utils/state_locks_filter.py

```diff
@@ -0,0 +1,391 @@
+#!/usr/bin/env python3
+
+import argparse
+import re
+import sys
+
+
+class LockOpTrace:
+    """
+    LockOpTrace: Represents a Lock Operation trace in the log file.
+    It holds the lines for the trace and allows matching against a part of the trace.
+    """
+
+    def __init__(self, lines: list[str]):
+        self.lines = lines
+        self.hash = hash(str(self))
+
+    def get_trace_lines(self) -> list[str]:
+        return self.lines
+
+    def match(self, part: str) -> bool:
+        for line in self.lines:
+            if part in line:
+                return True
+
+        return False
+
+    def __str__(self) -> str:
+        return "".join(self.lines)
+
+    def __hash__(self):
+        return self.hash
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, LockOpTrace):
+            # don't attempt to compare against unrelated types
+            return NotImplemented
+
+        return hash(str(self)) == hash(str(other))
```

Comment:
I think this should first check the hash and then check the contents of lines.

Reply:
On top of that this undoes the optimization you did on __hash__.

Lastly, in general, __eq__ should really not check for hashes of the string representation of objects that may be distinct for other reasons.